[FIX] web_editor: hide ghost overlay

rahg-odoo opened the pull request:
Steps to reproduce:
- Add an snippet.
- Click the "show/hide on desktop"  button of visibility option.
- Click on the little eye on the "hidden elements" section to have it displayed again.
- Go to mobile preview mode.
- Come back to desktop view by toggling off mobile preview button.
- An empty overlay appears on the left of the screen. With dimensions similar to snippet in mobile view.

After this commit:
Empty overlays no longer appear when switching between preview modes.

task-3270034

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/173328.png)](https://mergebot.odoo.com/odoo/odoo/pull/173328)

shsa-odoo commented:
@rahg-odoo Commit message and PR message missing, explain what was the issue and what changes you introduced  in that message. Thanks !

shsa-odoo commented on addons/website/static/src/js/editor/snippets.options.js:
```diff
@@ -2614,6 +2624,12 @@ options.registry.DeviceVisibility = options.Class.extend({
             this.$target[0].classList.add('o_snippet_override_invisible');
         }
+        if(this.$target[0].classList.contains('o_snippet_desktop_invisible')){
+            this.$target[0].dataset.tempDesktopPreview = "1"
```
semicolon missing here!

shsa-odoo commented on addons/website/static/src/js/editor/snippets.options.js:
```diff
@@ -2614,6 +2624,12 @@ options.registry.DeviceVisibility = options.Class.extend({
+            this.$target[0].dataset.tempDesktopPreview = "1"
+        }
+        if(this.$target[0].classList.contains('o_snippet_mobile_invisible')){
+            this.$target[0].dataset.tempMobilePreview = "1"
```
same here!

shsa-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -4546,6 +4546,10 @@ var SnippetsMenu = Widget.extend({
                 if (isMobilePreview === isMobileHidden) {
                     invisibleOverrideEl.dataset.invisible = '1';
+                    if(invisibleOverrideEl.dataset.tempMobilePreview === "1" || invisibleOverrideEl.dataset.tempDesktopPreview === "1"){
+                        invisibleOverrideEl.classList.add('o_snippet_override_invisible');
```
use double quotes.

shsa-odoo commented on addons/website/static/src/js/editor/snippets.options.js:
```diff
@@ -2600,6 +2604,12 @@ options.registry.DeviceVisibility = options.Class.extend({
      */
     async onTargetHide() {
         this.$target[0].classList.remove('o_snippet_override_invisible');
+        if(this.$target[0].classList.contains('o_snippet_desktop_invisible')){
+             delete this.$target[0].dataset.tempDesktopPreview;
+        }
+        if(this.$target[0].classList.contains('o_snippet_mobile_invisible')){
```
use double quotes.

rahg-odoo commented:
hello @shsa-odoo,

changes with new approach and original specs are pushed. that's why writing commit message that early was not good idea :smile: Thanks!!

shsa-odoo commented:
> hello @shsa-odoo,
> changes with new approach and original specs are pushed. that's why writing commit message that early was not good idea 😄 Thanks!!

@rahg-odoo True! make sure send task to review once spec is properly implemented, makes it easier to review :wink: , also looking at your current fix, it looks good, Good Work!

msh-odoo reviewed (commented):
@rahg-odoo As we are triggerring ```snippet_option_visibility_update``` then calling method 
```this._updateInvisibleDOM();``` at line 4566 is useless I guess.

rahg-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -4554,6 +4554,8 @@ var SnippetsMenu = Widget.extend({
                 invisibleOverrideEl.classList.remove('o_snippet_override_invisible');
                 if (isMobilePreview === isMobileHidden) {
+                    // Updating options upon changing preview mode to avoid ghost overlay
+                    this.trigger_up("snippet_option_visibility_update", {show: false});
```
Done with both required chanegs!
Thank you ! : )

bvr-odoo reviewed (changes requested):
Hi @rahg-odoo 
Before I review the code:
- PR title => remove website, your change is only in web_editor.
- There are 2 commits in your PR with the same title. I guess something is wrong.
- Add the clear "steps to reproduce" in your commit/PR message (like for all fix commits)

rahg-odoo commented:
> Hi @rahg-odoo Before I review the code:
> 
> * PR title => remove website, your change is only in web_editor.
> * There are 2 commits in your PR with the same title. I guess something is wrong.
> * Add the clear "steps to reproduce" in your commit/PR message (like for all fix commits)

sorry for that! 
@bvr-odoo  Ready for review.

bvr-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -4560,6 +4560,15 @@ var SnippetsMenu = Widget.extend({
+                });
+            if (invisibleSnippetSection) {
+                this._activateSnippet(false);
+            }
```
What you do here is: if an element is invisible on the page, it deactivates the snippet.
This is wrong, your change introduces a worse bug 😬 . It deactivates any selected snippet (even not invisible one) when switching to desktop view if there is an invisible element on the page.
See video:

https://github.com/user-attachments/assets/21d084d5-88fd-444b-9b99-1d584b440682

Also I don't think you need to find again invisible elements, it's already done a few lines above (line 4553)

rahg-odoo replied:
@bvr-odoo I have pushed changes! It meets expected behaviour shown in video. I have used getSelection method to recognize currently selected element and compares it with all the invisible element present in DOM then refresh options accordingly.
Thank you!

bvr-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -4550,10 +4550,16 @@ var SnippetsMenu = Widget.extend({
             const mobileViewThreshold = MEDIAS_BREAKPOINTS[SIZES.LG].minWidth;
             const isMobilePreview = this.$body[0].ownerDocument.defaultView.frameElement.clientWidth < mobileViewThreshold;
+            const selection = this.options.wysiwyg.odooEditor.document.getSelection();
+            const selectedSection = selection.getRangeAt(0).startContainer.parentElement.closest("section");
```
The document selection is not always the snippet with the overlay. For example, in this screenshot:

![image](https://github.com/user-attachments/assets/53bf1710-eea3-48d4-adc9-94dbf1f50bcc)

The selection includes the text "Useful Links" which is below the selected snippet. In this case, the fix doesn’t work. I think we need to find another way to get the selected snippet.

rahg-odoo replied:
hello @bvr-odoo 
```const selectedSection = this.options.wysiwyg?.lastElement.closest("section");```  this one seems to be working with last suggestion when selection and overlay is different, ```getSelection()``` does not give always the correct answer for our use.
although I was just informing, test when it's open to review.

msh-odoo replied:
> The document selection is not always the snippet with the overlay. For example, in this screenshot:
> 
> ![image](https://private-user-images.githubusercontent.com/52911687/382756544-53bf1710-eea3-48d4-adc9-94dbf1f50bcc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzMDU2ODQsIm5iZiI6MTczMTMwNTM4NCwicGF0aCI6Ii81MjkxMTY4Ny8zODI3NTY1NDQtNTNiZjE3MTAtZWVhMy00OGQ0LWFkYzktOTRkYmYxZjUwYmNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTExVDA2MDk0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg4ZTEwYjU4MWQwZTNiYTZhNTg5NjI2ZWNjY2ZjYzg5YzBlZjZhOWNkNzc5NWE3Yzk0NWJlZmM5OTQ2MmJiNDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4p93fqmO_2FQuxpn69ux55K8eQ2pFAQBMR9sWcNNMcI)
> 
> The selection includes the text "Useful Links" which is below the selected snippet. In this case, the fix doesn’t work. I think we need to find another way to get the selected snippet.

@bvr-odoo This looks issue in Mobile, If we are selecting some text then current Snippet should be unselected, for instance the solution given by @rahg-odoo will work here but it looks issue in mobile screen/preview.

bvr-odoo reviewed (changes requested):
Hello @rahg-odoo 
It's still not the right fix. Please test your fix thoroughly before requesting a review.

msh-odoo commented:
> const selection = this.options.wysiwyg.odooEditor.document.getSelection();
>             const selectedSection = selectio

@bvr-odoo According to me this is bug in mobile preview or mobile screen, what I suggest is to keep the previous solution and fix this issue, when someone select something in mobile then it should deactivate the previously selected snippet.

rahg-odoo commented:
@msh-odoo right now, we can go with ```this.options.wysiwyg.lastElement.closest('.o_snippet_mobile_invisible, .o_snippet_desktop_invisible``` instead of  ```this.options.wysiwyg.lastElement.closest("section");``` to counter the problem when overlay is around non section element.

bvr-odoo commented:
Hi @rahg-odoo If this one is ready, could you please rebase it and ping me when it's green on runbot ? thanks

rahg-odoo commented:
> Hi @rahg-odoo If this one is ready, could you please rebase it and ping me when it's green on runbot ? thanks

@bvr-odoo Good to go. Thanks for review!

bvr-odoo reviewed (changes requested):
@rahg-odoo
Please 🙏🏼 , test your work before requesting a review 😬
I tested the steps to reproduce of your PR on the runbot (latest push)
And at the second step:
![image](https://github.com/user-attachments/assets/ab0e2945-667b-48e8-91e9-de4d5f74a5ca)

rahg-odoo commented:
> @rahg-odoo Please 🙏🏼 , test your work before requesting a review 😬 I tested the steps to reproduce of your PR on the runbot (latest push) And at the second step.

I have tested this multiple times, and no tracebacks appeared. Here’s a video demonstrating the steps I followed: https://www.loom.com/share/80c98488acdd489c9e98d052fb3f30a4?sid=db4f551c-520a-4e40-a60f-a3bf0436a313.

Steps:

Added the snippet.
Toggled off desktop view.
Clicked on the eye button.
Switched the preview to mobile and then back to desktop.

Although I couldn’t reproduce the issue, I acknowledge that a traceback occurred on your end. To prevent potential errors, I am adding the optional chaining (?.) operator after lastElement to avoid calling closest on undefined.

rahg-odoo commented:
> Hello @rahg-odoo I understood the step `Toggle off the desktop view button` as clicking on this button.

Hi @bvr-odoo changed commit message and traceback is handled. do let me know if anything else goes wrong.(hope not!)
thanks for review and insights!

bvr-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -4568,10 +4568,19 @@ var SnippetsMenu = Widget.extend({
+                    // Updating options upon changing preview mode to avoid ghost overlay
+                    if (invisibleOverrideEl === withOverlayEl) {
+                        this.trigger_up("snippet_option_visibility_update", {show: false});
+                    }
```
Is it not simply:

- If the selected snippet becomes invisible after clicking the mobile/device preview button, then deactivate the snippet.

So in the end, if the selected snippet has `dataset.invisible = 1`, deactivate it.

Anyway, the code is incomplete as it is now. It works when the snippet is hidden on desktop, but not the other way -> when it is hidden on mobile.

Try this:

- Drop a snippet.
- Hide it on mobile.
- Switch to mobile preview.
- The overlay is still there (if not move your mouse everywhere), and the options are still enabled.

![image](https://github.com/user-attachments/assets/ade01a7e-7e65-48be-8304-6b77b8f72f76)

bvr-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -4577,14 +4577,29 @@ var SnippetsMenu = Widget.extend({
+            );
+            const needDeactivate = enabledInvisibleOverrideEl && enabledInvisibleOverrideEl.dataset.invisible === "1";
+            if (needDeactivate) {
+                this._activateSnippet(false);
```
Why is it done here and not before the `resolve()`? I ask because it could avoid doing `if (needDeactivate)` twice.

rahg-odoo replied:
I did it because `_updateInvisibleDOM` was below `_activateSnippet` in `snippet_option_visibility_update` event which used before, but it doesn't make any difference so pushed new changes. Thanks for insights and review!

bvr-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -4581,10 +4581,21 @@ var SnippetsMenu = Widget.extend({
 
+            // Updating options upon changing preview mode to avoid ghost overlay
+            const enabledInvisibleOverrideEl = this.options.wysiwyg.lastElement &&
+            this.options.wysiwyg.lastElement.closest(".o_snippet_mobile_invisible, .o_snippet_desktop_invisible");
```
once again => need to be indented

msh-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -4577,14 +4577,29 @@ var SnippetsMenu = Widget.extend({
+            // Updating options upon changing preview mode to avoid ghost overlay
+            const enabledInvisibleOverrideEl =
+            this.options.wysiwyg.lastElement &&
+            this.options.wysiwyg.lastElement.closest(
```
BTW, we follow the Odoo tooling for linting, I guess @rahg-odoo might have followed the suggestion given by linter.

bvr-odoo reviewed (approved):
LGTM 👍🏼

bvr-odoo commented:
@robodoo r+
